gno.land/cmd/gnoland/secrets_get.go

```diff
@@ -68,63 +78,161 @@ func execSecretsGet(cfg *secretsGetCfg, args []string, io commands.IO) error {
 		nodeKeyPath        = filepath.Join(cfg.dataDir, defaultNodeKeyName)
 	)
 
+	// Construct the display methods
+	var (
+		displayVK = wrapDisplayFn(io, outputTerminalVK)
+		displayVS = wrapDisplayFn(io, outputTerminalVS)
+		displayNK = wrapDisplayFn(io, outputTerminalNK)
+	)
+
+	if cfg.json {
+		displayVK = wrapDisplayFn(io, outputJSONCommon[validatorKeyInfo])
+		displayVS = wrapDisplayFn(io, outputJSONCommon[validatorStateInfo])
+		displayNK = wrapDisplayFn(io, outputJSONCommon[nodeKeyInfo])
+	}
+
 	switch key {
 	case validatorPrivateKeyKey:
 		// Show the validator's key info
-		return readAndShowValidatorKey(validatorKeyPath, io)
+		return readAndShowValidatorKey(validatorKeyPath, displayVK)
 	case validatorStateKey:
 		// Show the validator's last sign state
-		return readAndShowValidatorState(validatorStatePath, io)
+		return readAndShowValidatorState(validatorStatePath, displayVS)
 	case nodeKeyKey:
 		// Show the node's p2p info
-		return readAndShowNodeKey(nodeKeyPath, io)
+		return readAndShowNodeKey(nodeKeyPath, displayNK)
 	default:
 		// Show the node's p2p info
-		if err := readAndShowNodeKey(nodeKeyPath, io); err != nil {
+		if err := readAndShowNodeKey(nodeKeyPath, displayNK); err != nil {
 			return err
 		}
 
 		// Show the validator's key info
-		if err := readAndShowValidatorKey(validatorKeyPath, io); err != nil {
+		if err := readAndShowValidatorKey(validatorKeyPath, displayVK); err != nil {
 			return err
 		}
 
 		// Show the validator's last sign state
-		return readAndShowValidatorState(validatorStatePath, io)
+		return readAndShowValidatorState(validatorStatePath, displayVS)
+	}
+}
+
+type (
+	validatorKeyInfo struct {
+		Address string `json:"address"`
+		PubKey  string `json:"pub_key"`
+	}
+
+	validatorStateInfo struct {
+		Height int64 `json:"height"`
+		Round  int   `json:"round"`
+		Step   int8  `json:"step"`
+
+		Signature []byte `json:"signature"`
+		SignBytes []byte `json:"sign_bytes"`
+	}
+
+	nodeKeyInfo struct {
+		NodeID string `json:"node_id"`
+	}
+)
+
+func (v validatorStateInfo) MarshalJSON() ([]byte, error) {
+	type original validatorStateInfo
+
+	return json.Marshal(&struct {
+		Signature string `json:"signature,omitempty"`
+		SignBytes string `json:"sign_bytes,omitempty"`
+		original
+	}{
+		Signature: fmt.Sprintf("%X", v.Signature),
+		SignBytes: fmt.Sprintf("%X", v.SignBytes),
+		original:  (original)(v),
+	})
+}
+
+type (
+	secretDisplayType interface {
+		validatorKeyInfo | validatorStateInfo | nodeKeyInfo
+	}
+
+	displayFn[T secretDisplayType] func(input T) error
+	outputFn[T secretDisplayType]  func(input T, io commands.IO) error
+)
```

Comment:
I'm probably too old-school, but I'm unsure of the purpose of using a generic approach instead of directly interfacing each structure with a marshal, display, or similar method.

Reply:
Agree with @gfanton , I think we can simplify this logic a bit by using a common interface in a way that processes the data in a specific way depending on the implementation.